### PR TITLE
Fix for #6855 - Measure-Object should handle scriptblock properties.

### PR DESCRIPTION
**Pull request by BrucePay**
## PR Summary

Fixed issue #6855 by renaming the class that does all the work from `MshExpression` to `PSPropertyExpression`, making the class public and then in `MeasureObjectCommand`, lifting it up to the parameter level. Previously the implementation exposed the `-Property` as `string` then wrapped it internally as a `PSPropertyExpression`. Now the parameter type is `PSPropertyExpression` directly allowing for both wildcard strings and scriptblocks.  Note that this is technically a breaking change (the public type of the parameter has changed) but this was reviewed by the committee and approved. (The vast majority of changes in this PR are due to the renaming. The code changes are relatively small.)

`PSPropertyExpression` now lives in a public namespace where it can be used by cmdlet and script authors to easily add the same type of functionality to their commands. 

I also modified `PSPropertyExpression` to treat hashtables as objects so that
```
@{prop = 3} | measure-object prop
```
and
```
@{prop = 3} | measure-object {$_.prop}
```
will work the identically. (Previously the example using just the property name would just fail.)

#### Breaking Change Summary

This PR brings in a breaking change in an edge case:
> This could be a breaking change when it comes to the ConstrainedLanguage mode.
@{prop = 3} | measure-object Count works in ConstrainedLanguage mode today, but with this change, the conversion from Hashtable to PSCustomObject will throw exception.

It has been reviewed by the committee and agreed to accept this breaking change.
Please see the discussion at https://github.com/PowerShell/PowerShell/pull/6934#discussion_r193243793 for more information.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
@BrucePay Thanks! I am extecting that the PR will marked by the label.
Seems we lost all MSFT team in recent weeks. I wonder.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1]: Fixed.

**Comment by BrucePay** [comment]
Oops - I had the wrong issue number in the comment so you might have looked at the wrong issue. #6855 is, in fact, marked "Committee Reviewed".

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1]: Moved into a separate Describe.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1]: Added.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: Could you please use C# 7 syntax?
```c#
if (PSObject.Base(target) as Hashtable targetAsHash) 
```

**Review by iSazonov — COMMENTED** [review]
LGTM with one minor comment.

**Review by iSazonov — APPROVED** [review]

**Comment by BrucePay** [comment]
@SteveL-MSFT Says we're ignoring CodeFactor issues for now.

**Comment by SteveL-MSFT** [comment]
@iSazonov FYI - Most of my team has been involved with helping port existing Windows PowerShell modules owned by other teams to work with PSCore6 so you may find us not as active.  Hopefully we'll be back to fully engaged by end of June (or mostly engaged).

**Comment by iSazonov** [comment]
I'm glad that this process began. I do not see this leading to corrections in PowerShell Core - it convinces me that PowerShell Core is very good! :smile:

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: This could be a breaking change when it comes to the `ConstrainedLanguage` mode.
`@{prop = 3} | measure-object Count` works in `ConstrainedLanguage` mode today, but with this change, the conversion from `Hashtable` to `PSCustomObject` will throw exception.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: Can you please change the name to `WrapHashtableAsPSCustomObject`? C# method naming conversion suggests using verbs or verb phrases to name methods.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: If we want to handle hashtable specially, shall we expand that to `IDictionary`? `LanguagePrimitives.ConvertPSObjectToType` is able to convert `IDictionary` to `PSObject`.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
/cc @PaulHigin Now that `PSPropertyExpression` is publically exposed and it has the ability to run scripts, Paul you may want to check if this will introduce any security concerns to the `ConstrainedLanguage` mode.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (BrucePay) [reply src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: The point of the rule is that the names of methods should be imperative which this is. There are other examples of (type 1) conditional clauses in .NET such as  `ThrowIfCancellationRequested` on `System.Threading.CancellationToken`.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (BrucePay) [reply src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: The goal is to match the behaviour of `[PSCustomObject] @{a=1; b=2}`. Currently casting `Dictionary<T,T>` to `PSCustomObject` doesn't result in  a custom object so I don't think we should introduce that behaviour here.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (BrucePay) [reply src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: Hmm - true - good catch. It's pretty obscure and currently the only useful scenario for piping hashtables into `measure-object` (other than instance-counting). Still, we should probably get a committee review on this.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: Understood. Good to know that we are matching the behavior of `[PSCustomObject] @{a=1; b=2}`.

File: src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs]: OK. #Closed

**Review by daxian-dbw — APPROVED** [review]
Waiting on the committee's review.

**Comment by PaulHigin** [comment]
@daxian-dbw @BrucePay  Sorry for the late review.  To answer Dongbo's question, from a security point of view I think this is Ok.  MshExpression (now PSPropertyExpression) takes a ScriptBlock in its constructor which it will invoke.  But as long as the passed in ScriptBlock is created in a constrained environment it will be marked accordingly and the script block will be invoked with CL mode constraints.  

A constructor also takes a string "expression", but as near as I can tell it is never evaluated as script and is used to look up properties.  Is this correct?

However, we should include a CL mode test to ensure the expression scriptblock is run in CL mode.

**Comment by daxian-dbw** [comment]
> A constructor also takes a string "expression", but as near as I can tell it is never evaluated as script and is used to look up properties. Is this correct?

@PaulHigin Yes, that's correct. It's for property lookup.

File: src/System.Management.Automation/engine/parser/TypeResolver.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/TypeResolver.cs]: @PaulHigin I would like you to take another look at this change. `PSPropertyExpression` is added to `CoreTypes`, and that means users can call its methods in `ConstrainedLanguage` mode. Do you have any concerns about it?

Comment (PaulHigin) [reply src/System.Management.Automation/engine/parser/TypeResolver.cs]: Thanks, I didn't see that.  If this type is going to be white listed then it needs to go through a separate/full security review.  I see that some internal methods are now public, and these along with constructors need to be reviewed for vulnerabilities.  Please set up a meeting for this.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is fine with this breaking change which is likely only an edge case, but the positive impact for most users is beneficial

File: src/System.Management.Automation/engine/parser/TypeResolver.cs
Comment (BrucePay) [reply src/System.Management.Automation/engine/parser/TypeResolver.cs]: No - it's only intended to be a type accelerator. I added it in the wrong place. I'll move it to the `TypeAccelerators()` method where it should be.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/parser/TypeResolver.cs]: Yeah, I agree it should be moved :)

**Comment by daxian-dbw** [comment]
@PaulHigin New commits have been pushed. Can you please update your review? Thanks

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
@BrucePay I just noticed that we didn't run feature tests for this PR CI. Can you please push another commit with `[feature]` in the commit message? That will trigger the feature test run. You can also update the current head commit by running `git commit --amend` in your local `brucepay_MeasureObject` branch, and then forcing push to the remote by `git push --force-with-lease <your fork repo> brucepay_MeasureObject`. Thanks!

**Comment by daxian-dbw** [comment]
@BrucePay I have submitted a dummy commit to trigger the feature test run for this PR.

**Pull request merged**
